File: sdk/python/tempmail_sdk/providers/gonebox_email.py
"""
gonebox-email 渠道 — https://gonebox.email

一次性临时邮箱服务，无需认证。
  1. 创建: POST https://api.gonebox.email/api/v1/inboxes body: {"domain":"gonebox.email"}
     响应: {"success":true,"data":{"id":"...","address":"...","domain":"...","expiresAt":...,"ttl":...}}
  2. 读信: GET https://api.gonebox.email/api/v1/inboxes/{address}/messages
     响应: {"success":true,"data":{"address":"...","count":0,"messages":[...]}}
"""

from typing import List

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "gonebox-email"
BASE_URL = "https://api.gonebox.email/api/v1"

_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36"
    ),
    "Accept": "application/json",
    "Content-Type": "application/json",
}


def generate_email() -> EmailInfo:
    """
    创建 gonebox.email 临时邮箱
    POST /inboxes 请求分配一个随机邮箱地址，无需认证
    """
    resp = tm_http.post(
        f"{BASE_URL}/inboxes",
        headers=_HEADERS,
        json={"domain": "gonebox.email"},
        timeout=15,
    )
    resp.raise_for_status()
    body = resp.json()

    if not isinstance(body, dict) or not body.get("success"):
        raise RuntimeError(f"gonebox-email: 创建邮箱失败，响应: {body!r}")

    data = body.get("data")
    if not isinstance(data, dict):
        raise RuntimeError(f"gonebox-email: 响应 data 格式无效，响应: {body!r}")

    address = data.get("address", "")
    if not address:
        raise RuntimeError(f"gonebox-email: 缺少邮箱地址，响应: {body!r}")

    # expiresAt 为秒级时间戳，转为毫秒
    expires_at = None
    raw_expires = data.get("expiresAt")
    if raw_expires:
        try:
            expires_at = int(raw_expires) * 1000
        except (ValueError, TypeError):
            pass

    return EmailInfo(
        channel=CHANNEL,
        email=address,
        _token="",
        expires_at=expires_at,
    )


def get_emails(token: str, email: str) -> List[Email]:
    """
    获取 gonebox.email 邮件列表
    GET /inboxes/{address}/messages，无需认证
    """
    address = (email or "").strip()
    if not address:
        raise ValueError("gonebox-email: 缺少邮箱地址")

    headers = {k: v for k, v in _HEADERS.items() if k != "Content-Type"}

    resp = tm_http.get(
        f"{BASE_URL}/inboxes/{address}/messages",
        headers=headers,
        timeout=15,
    )
    if resp.status_code == 404:
        return []
    resp.raise_for_status()

    body = resp.json()
    if not isinstance(body, dict) or not body.get("success"):
        return []

    data = body.get("data")
    if not isinstance(data, dict):
        return []

    messages = data.get("messages")
    if not isinstance(messages, list) or not messages:
        return []

    out: List[Email] = []
    for msg in messages:
        if not isinstance(msg, dict):
            continue
        raw = {
            "id": msg.get("id", ""),
            "from": msg.get("from", "") or msg.get("sender", ""),
            "to": address,
            "subject": msg.get("subject", ""),
            "text": msg.get("text", "") or msg.get("body_plain", ""),
            "html": msg.get("html", "") or msg.get("body_html", ""),
            "date": str(msg.get("date", "") or msg.get("received_at", "")),
            "is_read": False,
            "attachments": [],
        }
        out.append(normalize_email(raw, address))

    return out

File: sdk/ruby/lib/tempmail_sdk/providers/gonebox_email.rb
# frozen_string_literal: true

module TempmailSdk
  module Providers
    # gonebox.email 渠道实现
    # 一次性临时邮箱服务，无需认证
    module GoneboxEmail
      CHANNEL = "gonebox-email"
      BASE_URL = "https://api.gonebox.email/api/v1"
      HEADERS = {
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
                        "(KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36",
        "Accept" => "application/json",
        "Content-Type" => "application/json"
      }.freeze

      module_function

      # @return [EmailInfo]
      def generate_email
        resp = Http.post(
          "#{BASE_URL}/inboxes",
          headers: HEADERS,
          json: { "domain" => "gonebox.email" },
          timeout: 15
        )
        resp.raise_for_status
        body = resp.json
        raise "gonebox-email: create failed" unless body.is_a?(Hash) && body["success"]

        data = body["data"]
        raise "gonebox-email: invalid data" unless data.is_a?(Hash)

        address = data["address"].to_s
        raise "gonebox-email: missing address" if address.empty?

        # expiresAt 为秒级时间戳，转为毫秒
        expires_at = nil
        raw_expires = data["expiresAt"]
        if raw_expires
          begin
            expires_at = raw_expires.to_i * 1000
          rescue StandardError
            # 忽略
          end
        end

        EmailInfo.new(channel: CHANNEL, email: address, token: "", expires_at: expires_at)
      end

      # @param token [String]
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        address = email.to_s.strip
        raise "gonebox-email: missing email" if address.empty?

        headers = HEADERS.reject { |k, _| k == "Content-Type" }
        resp = Http.get("#{BASE_URL}/inboxes/#{address}/messages", headers: headers, timeout: 15)
        return [] if resp.status_code == 404

        resp.raise_for_status
        body = resp.json
        return [] unless body.is_a?(Hash) && body["success"]

        data = body["data"]
        return [] unless data.is_a?(Hash)

        messages = data["messages"]
        return [] unless messages.is_a?(Array) && !messages.empty?

        messages.select { |m| m.is_a?(Hash) }.map do |msg|
          raw = {
            "id" => msg["id"].to_s,
            "from" => (msg["from"] || msg["sender"]).to_s,
            "to" => address,
            "subject" => msg["subject"].to_s,
            "text" => (msg["text"] || msg["body_plain"]).to_s,
            "html" => (msg["html"] || msg["body_html"]).to_s,
            "date" => (msg["date"] || msg["received_at"]).to_s,
            "is_read" => false,
            "attachments" => []
          }
          Normalize.normalize_email(raw, address)
        end
      end
    end
  end
end
